### PR TITLE
Add ability to manually pass argument to composable viewModel factories

### PR DESCRIPTION
**Pull request by jakoss**
Added a `argsFactory` parameter to `mavericksViewModel` so user can manually pass argument to view model.

Fixes: https://github.com/airbnb/mavericks/issues/573

### CODE REVIEW

File: mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt [hunk 1]
```diff
@@ -61,16 +64,27 @@ inline fun <reified VM : MavericksViewModel<S>, reified S : MavericksState> mave
...
-        when (scope) {
-            is Fragment -> {
-                val args = scope.arguments?.get(Mavericks.KEY_ARG)
-                FragmentViewModelContext(activity, args, scope)
-            }
-            else -> {
-                val args = activity.intent.extras?.get(Mavericks.KEY_ARG)
-                ActivityViewModelContext(activity, args, viewModelStoreOwner, savedStateRegistry)
+        var view: View? = localView
+        var fragment: Fragment? = null
```
Comment (elihart) [inline mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt:71]: I think the try/catch approach is fine, and is maybe preferable to not rely on the implementation detail of the `fragment_container_view_tag` tag. Main thing I would say though is to extract this to a dedicated function and add some documentation and where this pattern was taken from (eg point to the androidx code)

File: sample-compose/src/main/java/com/airbnb/mvrx/compose/sample/ComposeSampleActivity.kt [hunk 2]
```diff
@@ -25,7 +27,32 @@ import com.airbnb.mvrx.compose.mavericksViewModel
 
 data class CounterState(
     val count: Int = 0,
-) : MavericksState
+) : MavericksState {
+    constructor(arguments: ArgumentsTest): this(count = arguments.count)
+}
+
+data class ArgumentsTest(val count: Int) : Parcelable {
```
Comment (elihart) [inline sample-compose/src/main/java/com/airbnb/mvrx/compose/sample/ComposeSampleActivity.kt:34]: if you're able to add any actual unit tests that would be great, so we can catch regressions automatically in the future

**Review by elihart — COMMENTED** [review]
Thanks @jakoss, the general approach looks good, just a few requests for cleaning it up

File: mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt (hunk 1 above)
Comment (jakoss) [reply mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt:71]: The issue here is that this whole factory is inline. So any function we want to extract have to be public in na namespace of mavericks

Comment (jakoss) [reply mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt:71]: I solved that by using `@InternalMavericksApi`, I hope that's enough

File: sample-compose/src/main/java/com/airbnb/mvrx/compose/sample/ComposeSampleActivity.kt (hunk 2 above)
Comment (jakoss) [reply sample-compose/src/main/java/com/airbnb/mvrx/compose/sample/ComposeSampleActivity.kt:34]: I can add unit tests, but we need whole setup for compose tests and that seems like a work for separate PR. Do you agree with that?

Comment (jakoss) [reply sample-compose/src/main/java/com/airbnb/mvrx/compose/sample/ComposeSampleActivity.kt:34]: I updated the test in sample-compose in this PR. I will work on more generic tests for mvrx-compose

**Comment by jakoss** [comment]
@elihart I've added a test setup for `mvrx-compose` with 2 initial tests - passing arguments via `argsFactory` and passing arguments from the `Fragment.argument`. This should make it for regression tests

File: mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt [hunk 3]
```diff
@@ -34,25 +39,16 @@ import kotlin.reflect.KProperty1
...
-    } else {
-        while (currentContext is ContextWrapper) {
-            if (currentContext is ComponentActivity) {
-                activity = currentContext
-                break
-            }
-            currentContext = currentContext.baseContext
-        }
-    }
+    val activity = extractActivityFromContext(LocalContext.current)
```
Comment (elihart) [inline mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt:51]: thanks

**Review by elihart — APPROVED** [review]
looks great, appreciate the contribution. I'll push a release now

File: mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt (hunk 1 above)
Comment (elihart) [reply mvrx-compose/src/main/kotlin/com/airbnb/mvrx/compose/MavericksComposeExtensions.kt]: @jakoss that's fine. in the future for inline functions you can also make them `internal` and annotate with `@PublishedApi`

**Pull request merged**